server/src/test/java/io/crate/execution/dml/IndexerTest.java

```diff
@@ -1005,4 +1004,42 @@ public void test_index_nested_array() throws Exception {
             {"x":10,"y":[[1,2],[3,4]]}"""
         );
     }
+
+    @Test
+    public void test_generated_column_can_refer_to_a_non_string_partitioned_by_column() throws Exception {
+        String partition = new PartitionName(new RelationName("doc", "t"), List.of("2")).asIndexName();
+        SQLExecutor executor = SQLExecutor.builder(clusterService)
+            .addPartitionedTable("""
+             CREATE TABLE t (
+                 a INT,
+                 parted INT CHECK (parted > 1),
+                 gen_from_parted INT as parted + 1
+             ) PARTITIONED BY (parted)
+             """
+            ).build();
+        DocTableInfo table = executor.resolveTableInfo("t");
+        Indexer indexer = new Indexer(
+            partition,
+            table,
+            new CoordinatorTxnCtx(executor.getSessionSettings()),
+            executor.nodeCtx,
+            column -> NumberFieldMapper.FIELD_TYPE,
+            List.of(
+                table.getReference(new ColumnIdent("a"))
+                // 'Parted' is not in targets to imitate insert-from-subquery behavior
+                //  which excludes partitioned columns from targets
+            ),
+            null
+        );
+
+        // Imitating problematic query
+        // insert into t (a, parted) select 1, 2
+        // We are inserting into partition 2, so b = 2.
+        ParsedDocument parsedDoc = indexer.index(item(1));
```

Comment:
Initially had an issue with reproducing this bug with unit test as it passed even without patch. 

Verified that it now fails on master version and passes with the patch.